src/imodmodel/models.py

```diff
@@ -88,11 +96,49 @@ class Contour(BaseModel):
     """https://bio3d.colorado.edu/imod/doc/binspec.html"""
     header: ContourHeader
     points: np.ndarray  # pt
+    extra: List[GeneralStorage] = []
 
     class Config:
         arbitrary_types_allowed = True
 
 
+class MeshHeader(BaseModel):
+    """https://bio3d.colorado.edu/imod/doc/binspec.html"""
+    vsize: int
+    lsize: int
+    flag: int
+    time: int
+    surf: int
+
+class Mesh(BaseModel):
+    """https://bio3d.colorado.edu/imod/doc/binspec.html"""
+    header: MeshHeader
+    vertices: np.ndarray # vert
+    indices: np.ndarray # list
+    extra: List[GeneralStorage] = []
+
+    class Config:
+        arbitrary_types_allowed = True
+
+    def reshaped_indices(self) -> np.ndarray:
+        return self.indices[np.where(self.indices >= 0)].reshape((-1, 3))
```

Comment:
model looks good here - rather than having a reshaped indices method (or a [property](https://www.tutorialsteacher.com/python/property-decorator), for cleaner syntax) I would probably implement a [`field_validator`](https://docs.pydantic.dev/latest/usage/validators/#field-validators) to instead ensure the field has the correct shape

Reply:
Yes a property is a very good idea. I could make the raw indices (and maybe also vertices) private variables and put the cleanup and reshaping into properties.

I've never used pydantic before, but that sounds like a plan.

Reply:
A field validator is a little more appropriate than a property here, it's basically where you put logic to ensure data in a field is parsed correctly (validated) so you can just use the field rather than adding another property - I vote let's go with a validator here 🙂

Reply:
I think a validator is a good idea, but doesn't solve the problem why I wrote the reshaped_indices method:
The indices in the file look like this: `[-25, 1, 2, 3, 4, .... -22, -1]`. All values >= 0 are actual indices, all values < 0 are special values:
```
  -1   end of list array                
  -20  next item on list is normal vector.
  -21  begin polygon with vertices and vertex indices only <--- what I'm doing right now and what Napari wants
  -22  end polygon
  -23  begin vertex,normal polygon pairs with normal,vertex indices
  -24  begin large convex polygon with normals
  -25  begin vertex,normal polygon pairs with vertex indices <--- the only type of mesh that I implemented so far
```
To get a "clean" indices list for Napari, I'm removing all values < 0 and reshape it to (-1, 3), so I'm storing vertex indices only. However, to correctly interpret a MEST chunk one needs the raw indices list with control values, that's why I'm storing it like that in the object and only create the "clean" version on demand.

Maybe the better way would be:
- validate the indices during parsing: check the shape, check that all mesh types are supported and so on
- store the validated list as-is
- write a function/property that produces the indices in the shape and form that one needs
  - vertex_indices
  - raw_indices
  - ...?

Reply:
wow... this format is a mess 😆 thank you for patiently explaining - in this case what you suggest makes perfect sense - store the data in the file then further parse somewhere downstream, a property for the simple case seems very reasonable